File: eventstore.js
var data = {};

module.exports = {
	
	loadHistoryForId : function (id) {
		
		return data[id] || [];
	},
	
	save : function (id, e, v) {
		
		if (!e)
			throw "Event is undefined"
		
		var events = module.exports.loadHistoryForId(id);
		
		if (events.length - 1 > v) {
			throw "YAY";
		}
		
		e.version = events.length;
		e.time = new Date().getTime();
		
		data[id] = events.concat([e]);
	}

}

File: CommandAndQueries/CommandAndQueries/EventStore.swift
//
//  EventStore.swift
//  CommandAndQueries
//
//  Created by Vitaly Baum on 11/10/14.
//  Copyright (c) 2014 Vitaly Baum. All rights reserved.
//

import Foundation

struct EventStore {
    
    static let _instance = EventStore()
    
    static func instance() -> EventStore {
        return _instance;
    }
    
    func loadHistoryFor(id: String) -> [Event] {
        
        return JSONHelper.getList(id)
        
    }
  
    func save<T: Event>(id: String, e:T, v: Int) {
        
        var events:[Event] = loadHistoryFor(id)
        
        if events.count - 1 != v {
            NSException(
                name: "YAY!",
                reason: "Concurrent exception is happen for \(events.count)",
                userInfo: nil
                ).raise()
        }
        
        e.version = events.count
        e.time = Int64( NSDate.timeIntervalSinceReferenceDate());
        
        events.append(e)
        
        JSONHelper.saveList(id, items: events)
    }
    
    struct JSONHelper {
        
        static var data = Dictionary<String, [Event]> ();
        
        static func getList(id: String) -> [Event] {
            return data[id] ?? [Event]()
        }
        
        
        static func saveList(id: String, items: [Event]) {
            
            data[id] = items
        }
        
    }
    
}
